### PR TITLE
Github action: add deploy workflow

### PR DESCRIPTION
**Pull request by junghans**
This will automatically deploy the page when pushing to source branch (i.e. `master` and `source`).
That basically allows editing through the GitHub web interface, no more need to install bundler locally.

This also supports pull requests, which can be useful to collaboratively make a conference page. For PRs the whole build is still done except the actual `git push` at the end. To achieve this I added a `--dry-run` option to `bin/deploy`.

Side notes:
- added `--verbose` option to make debugging of `bin/deploy` easier
- changed the exit code for unknown options to 1, before it would end gracefully (`exit 0`)
- fixed `--user` option calling `shift` twice
- using actions/cache to cache bundle
- the current branch is always used as the source branch, however the deploy action is only triggered for `master` / `source` (and PR to these, but PR are not deployed).
- in case of a `.github.io` repo, the deploy branch is switched from `gh-pages` to `master`.
- all the GitHub action stuff is pretty easy to customize
- all changes to `bin/deploy` are backward compatible

### CODE REVIEW

**Comment by alshedivat** [comment]
@junghans, this is awesome, thanks for the contribution!

Overall, LGTM. Before merging this PR, let me play with the config to make sure it takes care of all / most of use cases.

File: bin/deploy
Comment (alshedivat) [inline bin/deploy:23]: Why was this `shift` deleted? Wouldn't this introduce a bug?

File: .github/workflows/deploy.yml
Comment (alshedivat) [inline .github/workflows/deploy.yml:37]: Where is this email coming from? Does it correspond to a specific GitHub action?

File: bin/deploy
Comment (alshedivat) [inline bin/deploy:34]: Same here (and on line 37 below as well): it looks like some options above have `shift` while others don't. Why and what's the expected behavior?

File: .github/workflows/deploy.yml
Comment (alshedivat) [inline .github/workflows/deploy.yml:43]: Could you explain the logic of this step:

1) Do I understand correctly that on PR, it will set `SRC_BRANCH` to `${GITHUB_BASE_REF}` (which must be the PR target) and also set the `--dry-run` option on. Why is dry-run necessary? What is the intended behavior on PR?
2) On push, it sets `SRC_BRANCH` to the branch name the push was made to, correct?

**Review by alshedivat — COMMENTED** [review]
LGTM.

Could you please explain and document the logic of this workflow a bit more by answering a couple of questions I left inline? I'd be happy to merge after that. Thanks.

File: bin/deploy
Comment (junghans) [inline bin/deploy:44]: @alshedivat There is already a shift here.

File: bin/deploy
Comment (junghans) [reply bin/deploy:34]: Line 44 has one `shift` for each option already, hence there is only a `shift` instead of a `shift 2` for `--src` as well.

File: .github/workflows/deploy.yml
Comment (junghans) [reply .github/workflows/deploy.yml:37]: You can use any email here, `<some_number>+github-actions[bot]@users.noreply.github.com` is the Github Actions default email. See https://github.community/t/github-actions-bot-email-address/17204 for discussion.

File: .github/workflows/deploy.yml
Comment (junghans) [reply .github/workflows/deploy.yml:43]: Re:  1. I was actually mistaken here, `GITHUB_BASE_REF` is the target branch, so we need to use `GITHUB_HEAD_REF` here (see https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions). The `--dry-run` is needed, so that a PR from an internal branch doesn't get deployed. PR from a fork won't have write access to the repo, so there it won't be an issue.

Re: 2. Yes, the action is only triggered when pushed to `master` or `source` and then `GITHUB_REF` will be something like `refs/heads/master`. It is worth mentioning that, in the user/org case, a push to the `source` branch will trigger the deploy action onto master, but the then automatic commit to the `master` branch won't trigger another Github action execution. GitHub doesn't allow a push triggered by an action to trigger another action.

File: bin/deploy
Comment (junghans) [reply bin/deploy:23]: Try `bin/deploy --user --src master` to reproduce the issue in the original deploy script.

File: .github/workflows/deploy.yml
Comment (alshedivat) [reply .github/workflows/deploy.yml:43]: Thanks for explaining!

Re: 1. In the dry-run regime, I guess base vs. head branch doesn't matter. Just to double-check the logic, on internal PR merge, the action will get triggered twice: (1)  due to PR (which will be a dry-run) and (2) due to the resulting commit from the PR in the base branch (which will trigger the actual deploy), right?

**Review by alshedivat — APPROVED** [review]

File: .github/workflows/deploy.yml
Comment (junghans) [reply .github/workflows/deploy.yml:43]: Re 1. I still think it needs to be `GITHUB_HEAD_REF`, let's assume you have PR from `new_feature` to `master`. The action is supposed to test the commits on `new_feature` (`GITHUB_HEAD_REF`) and not `master` (`GITHUB_HEAD_REF`). And this PR won't trigger twice as a push to `new_feature` doesn't trigger an action, only an open PR will trigger an action (see the selection of branches in the top of the file `on: push: branches:`)

Comment (junghans) [reply .github/workflows/deploy.yml:43]: Plus I will submit a bunch of unrelated other changes as PR afterwards to test this ;-)

Comment (alshedivat) [reply .github/workflows/deploy.yml:43]: > Re 1. I still think it needs to be `GITHUB_HEAD_REF`, let's assume you have PR from `new_feature` to `master`. The action is supposed to test the commits on `new_feature` (`GITHUB_HEAD_REF`) and not `master` (`GITHUB_HEAD_REF`). And this PR won't trigger twice as a push to `new_feature` doesn't trigger an action, only an open PR will trigger an action (see the selection of branches in the top of the file `on: push: branches:`)

Got it, so the following adjustment should make it work as expected?
```suggestion
        if [[ ${GITHUB_REF} = refs/pull/*/merge ]]; then # pull request
          echo "::set-output name=SRC_BRANCH::${GITHUB_HEAD_REF}"
          echo "::set-output name=DRY_RUN::--dry-run"
        elif [[ ${GITHUB_REF} = refs/heads/* ]]; then # branch, e.g. master, source etc
          echo "::set-output name=SRC_BRANCH::${GITHUB_REF#refs/heads/}"
        fi
```

> Plus I will submit a bunch of unrelated other changes as PR afterwards to test this ;-)

Awesome, thanks! :)

**Pull request merged**
